src/request.ts

```diff
@@ -0,0 +1,126 @@
+import type { DocumentTypeDecoration } from "@graphql-typed-document-node/core";
+import { createSha256, extractOperationName, pruneObject } from "helpers";
+
+export type ModeFlags = "persisted" | "document" | "both";
+
+export type DocumentIdFn = <TResult, TVariables>(
+	query: DocumentTypeDecoration<TResult, TVariables>,
+) => string | undefined;
+
+export type GraphQLRequest<TVariables> = {
+	mode: ModeFlags;
+	operationName: string;
+	query: string | undefined;
+	documentId: string | undefined;
+	variables: TVariables | undefined;
+	extensions: Record<string, unknown>;
+};
+
+export const isPersistedQuery = <T>(request: GraphQLRequest<T>): boolean =>
+	request.mode === "persisted" || request.mode === "both";
+
+export const asQueryString = <TVariables>(
+	request: GraphQLRequest<TVariables>,
+) =>
+	new URLSearchParams(
+		pruneObject({
+			documentId: request.documentId,
+			query: request.query,
+			operationName: request.operationName,
+			variables: isNotEmpty(request.variables)
+				? JSON.stringify(request.variables)
+				: undefined,
+			extensions: isNotEmpty(request.extensions)
+				? JSON.stringify(request.extensions)
+				: undefined,
+		}),
+	);
+
+const isNotEmpty = (value: unknown) => value && Object.keys(value).length > 0;
+
+export const asRequestURL = <TVariables>(
+	url: string,
+	request: GraphQLRequest<TVariables>,
+): URL => {
+	const result = new URL(url);
+	const qs = asQueryString(request);
+	for (const [key, value] of qs) {
```

Comment:
Wondering if this is needed if asQueryString already prunes undefined keys. Could possibly just do `result.search =  asQueryString(request).toString()`